src/note/asset_base.rs

```diff
@@ -0,0 +1,230 @@
+};
+
+/// Note type identifier.
+#[derive(Clone, Copy, Debug, Eq)]
+pub struct AssetBase(pallas::Point);
+
+// AssetBase must implement PartialOrd and Ord to be used as a key in BTreeMap.
+impl PartialOrd for AssetBase {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for AssetBase {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_coord = self.0.to_affine().coordinates().unwrap();
+        let other_coord = other.0.to_affine().coordinates().unwrap();
+        self_coord
+            .x()
+            .cmp(other_coord.x())
+            .then_with(|| self_coord.y().cmp(other_coord.y()))
+    }
+}
+
+/// Personalization for the ZSA asset digest generator
+pub const ZSA_ASSET_DIGEST_PERSONALIZATION: &[u8; 16] = b"ZSA-Asset-Digest";
+
+///    AssetDigest for the ZSA asset
+///
+///    Defined in [ZIP-227: Issuance of Zcash Shielded Assets][assetdigest].
+///
+///    [assetdigest]: https://zips.z.cash/zip-0227.html#specification-asset-identifier-asset-digest-and-asset-base
+pub fn asset_digest(encode_asset_id: &[u8]) -> Blake2bHash {
+    Params::new()
+        .hash_length(64)
+        .personal(ZSA_ASSET_DIGEST_PERSONALIZATION)
+        .to_state()
+        .update(encode_asset_id)
+        .finalize()
+}
+
+/// Encoding the Asset Identifier, as defined in [ZIP 227][assetidentifier].
+///
+/// [assetidentifier]: https://zips.z.cash/zip-0227.html#specification-asset-identifier-asset-digest-and-asset-base
+pub fn encode_asset_id(
+    version: u8,
+    ik: &IssueValidatingKey<ZSASchnorr>,
+    asset_desc_hash: &[u8; 32],
+) -> Vec<u8> {
+    let ik_encoding = ik.encode();
+    let mut asset_id = Vec::with_capacity(1 + ik_encoding.len() + asset_desc_hash.len());
+    asset_id.push(version);
+    asset_id.extend(ik_encoding);
+    asset_id.extend_from_slice(&asset_desc_hash[..]);
+    asset_id
+}
+
+impl AssetBase {
+    /// Deserialize the AssetBase from a byte array.
+    pub fn from_bytes(bytes: &[u8; 32]) -> CtOption<Self> {
+        pallas::Point::from_bytes(bytes).map(AssetBase)
+    }
+
+    /// Serialize the AssetBase to its canonical byte representation.
+    pub fn to_bytes(self) -> [u8; 32] {
+        self.0.to_bytes()
+    }
+
+    /// Note type derivation.
+    ///
+    /// Defined in [ZIP-226: Transfer and Burn of Zcash Shielded Assets][assetbase].
+    ///
+    /// [assetbase]: https://zips.z.cash/zip-0226.html#asset-identifiers
+    ///
+    /// # Panics
+    ///
+    /// Panics if the derived AssetBase is the identity point.
+    #[allow(non_snake_case)]
+    pub fn derive(ik: &IssueValidatingKey<ZSASchnorr>, asset_desc_hash: &[u8; 32]) -> Self {
+        let version_byte: u8 = 0x00;
+
+        // EncodeAssetId(ik, asset_desc_hash) = version_byte || ik || asset_desc_hash
+        let asset_id = encode_asset_id(version_byte, ik, asset_desc_hash);
+        let asset_digest = asset_digest(&asset_id);
```

Comment:
This should be named `AssetBase::custom` rather than `AssetBase::derive`, to match the "Custom Asset" terminology from ZIP 227.

If my enum suggestion is taken:
```suggestion
    pub fn custom(asset_id: &AssetId) -> Self 
        let asset_digest = asset_id.asset_digest();
```